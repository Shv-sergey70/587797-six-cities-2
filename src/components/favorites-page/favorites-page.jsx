import React from 'react';
import PropTypes from 'prop-types';
import {connect} from "react-redux";
import PageHeader from '../page-header/page-header';
import FavoritesList from "../favorites-list/favorites-list";
import FavoritesEmpty from "../favorites-empty/favorites-empty";
import {Link} from "react-router-dom";
import Operation from "../../operation";
import offerPropType from '../../prop-types/offer';
import {Route} from "../../const/routes";

class FavoritesPage extends React.PureComponent {
  constructor(props) {
    super(props);

    this._isEmptyFavorites = this._isEmptyFavorites.bind(this);
  }

  componentDidMount() {
    const {
      loadFavorites
    } = this.props;

    loadFavorites();
  }

  _isEmptyFavorites() {
    const {
      favoriteOffers
    } = this.props;

    return favoriteOffers.length === 0;
  }

  render() {
    const {
      favoriteOffers
    } = this.props;

    return <div className={`page ${this._isEmptyFavorites ? `page--favorites-empty` : ``}`}>
      <PageHeader/>

      {this._isEmptyFavorites() ? <FavoritesEmpty/> : <FavoritesList
        favoriteOffers={favoriteOffers}
      />}

      <footer className="footer container">
        <Link to={Route.MAIN} className={`footer__logo-link`}>
          <img className="footer__logo" src="img/logo.svg" alt="6 cities logo" width="64" height="33"/>
        </Link>
      </footer>
    </div>;
  }
}

FavoritesPage.propTypes = {
  favoriteOffers: PropTypes.arrayOf(offerPropType),
  loadFavorites: PropTypes.func.isRequired
};

const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  favoriteOffers: state.favoriteOffers
});

const mapDispatchToProps = (dispatch) => ({
  loadFavorites: () => dispatch(Operation.loadFavorites())
});

export {FavoritesPage};

export default connect(mapStateToProps, mapDispatchToProps)(FavoritesPage);
