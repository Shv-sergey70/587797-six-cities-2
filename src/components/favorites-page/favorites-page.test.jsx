import React from 'react';
import renderer from 'react-test-renderer';
import {FavoritesPage} from "./favorites-page";
import {BrowserRouter as Router} from "react-router-dom";

jest.mock(`../page-header/page-header.jsx`, () => `PageHeader`);
jest.mock(`../favorites-empty/favorites-empty.jsx`, () => `FavoritesEmpty`);
jest.mock(`../favorites-list/favorites-list.jsx`, () => `FavoritesList`);

const offersData = [
  {
    id: 1,
    title: `Beautiful & luxurious apartment at great location`,
    previewImage: `/img/apartment-01.jpg`,
    isPremium: true,
    isFavorite: true,
    type: `apartment`,
    price: 80,
    rating: 4.4,
    bedrooms: 2,
    maxAdults: 2,
    images: [`/img/room.jpg`, `/img/apartment-01.jpg`, `/img/studio-01.jpg`],
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`, `Dishwasher`],
    description: `A quiet cozy and picturesque that hides behind a a river`
  },
  {
    id: 2,
    title: `Beautiful & luxurious apartment at great location2`,
    previewImage: `/img/apartment-01.jpg`,
    isPremium: false,
    isFavorite: true,
    type: `apartment`,
    price: 180,
    rating: 2.4,
    bedrooms: 5,
    maxAdults: 3,
    images: [`/img/room.jpg`, `/img/apartment-01.jpg`, `/img/studio-01.jpg`],
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`, `Coffee machine`],
    description: `A quiet cozy and picturesque that hides behind a a river1`
  },
  {
    id: 3,
    title: `Beautiful & luxurious apartment at great location33`,
    previewImage: `/img/apartment-01.jpg`,
    isPremium: true,
    isFavorite: false,
    type: `room`,
    price: 820,
    rating: 4.9,
    bedrooms: 5,
    maxAdults: 4,
    images: [`/img/room.jpg`, `/img/apartment-01.jpg`],
    goods: [`Heating`, `Kitchen`, `Cable TV`, `Washing machine`],
    description: `A quiet cozy and picturesque that hides behind a a river22`
  }
];

describe(`Favorites page behavior`, () => {
  it(`No favorites offers`, () => {
    const tree = renderer.create(
        <Router>
          <FavoritesPage
            favoriteOffers={[]}
            loadFavorites={jest.fn()}
          />
        </Router>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`With favorites offers`, () => {
    const tree = renderer.create(
        <Router>
          <FavoritesPage
            favoriteOffers={offersData}
            loadFavorites={jest.fn()}
          />
        </Router>).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
