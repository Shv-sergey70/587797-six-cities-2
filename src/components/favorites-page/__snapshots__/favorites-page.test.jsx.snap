// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites page behavior No favorites offers 1`] = `
<div
  className="page page--favorites-empty"
>
  <PageHeader />
  <FavoritesEmpty />
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;

exports[`Favorites page behavior With favorites offers 1`] = `
<div
  className="page page--favorites-empty"
>
  <PageHeader />
  <FavoritesList
    favoriteOffers={
      Array [
        Object {
          "bedrooms": 2,
          "description": "A quiet cozy and picturesque that hides behind a a river",
          "goods": Array [
            "Heating",
            "Kitchen",
            "Cable TV",
            "Washing machine",
            "Coffee machine",
            "Dishwasher",
          ],
          "id": 1,
          "images": Array [
            "/img/room.jpg",
            "/img/apartment-01.jpg",
            "/img/studio-01.jpg",
          ],
          "isFavorite": true,
          "isPremium": true,
          "maxAdults": 2,
          "previewImage": "/img/apartment-01.jpg",
          "price": 80,
          "rating": 4.4,
          "title": "Beautiful & luxurious apartment at great location",
          "type": "apartment",
        },
        Object {
          "bedrooms": 5,
          "description": "A quiet cozy and picturesque that hides behind a a river1",
          "goods": Array [
            "Heating",
            "Kitchen",
            "Cable TV",
            "Washing machine",
            "Coffee machine",
          ],
          "id": 2,
          "images": Array [
            "/img/room.jpg",
            "/img/apartment-01.jpg",
            "/img/studio-01.jpg",
          ],
          "isFavorite": true,
          "isPremium": false,
          "maxAdults": 3,
          "previewImage": "/img/apartment-01.jpg",
          "price": 180,
          "rating": 2.4,
          "title": "Beautiful & luxurious apartment at great location2",
          "type": "apartment",
        },
        Object {
          "bedrooms": 5,
          "description": "A quiet cozy and picturesque that hides behind a a river22",
          "goods": Array [
            "Heating",
            "Kitchen",
            "Cable TV",
            "Washing machine",
          ],
          "id": 3,
          "images": Array [
            "/img/room.jpg",
            "/img/apartment-01.jpg",
          ],
          "isFavorite": false,
          "isPremium": true,
          "maxAdults": 4,
          "previewImage": "/img/apartment-01.jpg",
          "price": 820,
          "rating": 4.9,
          "title": "Beautiful & luxurious apartment at great location33",
          "type": "room",
        },
      ]
    }
  />
  <footer
    className="footer container"
  >
    <a
      className="footer__logo-link"
      href="/"
      onClick={[Function]}
    >
      <img
        alt="6 cities logo"
        className="footer__logo"
        height="33"
        src="img/logo.svg"
        width="64"
      />
    </a>
  </footer>
</div>
`;
